### DISCUSSION TITLE
Flow diagrams: Zero-value and fee outputs

### DISCUSSION THREAD

**Discussion by casey** [mempool/mempool#2604] [General]
I love the flow diagrams! There are a couple of things that I think are a bit confusing though:

1. Zero-fee outputs appear in flow diagrams. For example, here's a coinbase output with three OP_RETURNS: <img width="1145" alt="Screen Shot 2022-10-04 at 10 42 56 AM" src="https://user-images.githubusercontent.com/1945/193889923-d2040329-4513-4134-a197-4c41b870da54.png">
  Conceptually, I think of the flow-diagram as tracking the flow of bitcoin, which zero-fee outputs don't participate in, so this is a little unintuitive to me.

2. The fee is the first output. For example, here's a normal transaction that pays a fee: <img width="1134" alt="Screen Shot 2022-10-04 at 10 48 17 AM" src="https://user-images.githubusercontent.com/1945/193890161-e094d0b8-0c5a-4306-847b-17cede30c4c6.png">
  Conceptually, I think of the fee as being what's left after deducting the outputs, so I think about it as being the last output, not the first.

These are matters of taste, so I wanted to open a discussion instead of an issue. What do y'all think about these aspects of flow diagrams?

**Comment by mononaut** [comment]
Great points!

(1) I can definitely see the case for omitting zero-value outputs from the diagram. I think it comes down to whether we want a precise graph of the flow of value, or a slightly fuzzier representation of the 'shape' of the transaction.

In the latter case, it makes sense to visually differentiate similar transactions with and without OP_RETURNS, even if those outputs aren't relevant to the value flow.

Perhaps a good compromise would be to replace the zero-value flow lines with a disconnected shape/label, so that there's still some differentiation, but the flow of value isn't confused. Something like:

![zero-value](https://user-images.githubusercontent.com/83316221/193921070-71f25f3b-009d-4dd6-86f9-d3d42061545d.png)

(2) The reasoning behind this choice is that the fee paid is often a more salient feature than the Nth output of a large transaction, so I wanted to keep the fee line "above the fold", as it were.

![d31c2b4d71eb143b23bb87919dda7fdfecee337ffa1468d1c431ece37698f918](https://user-images.githubusercontent.com/83316221/193921297-a0480cf8-86d0-49b5-80fa-c442c4f2d435.png)

But the conceptualization of the fee as "what's left" after the outputs is an excellent point I hadn't considered, and I suppose the fee is already highlighted prominently elsewhere on the page.

**Reply by casey** [reply]
Thanks for the through response!

1. Nice, I really like the way that looks. It shows zero-value outputs so you don't wonder where they are, but they don't participate in the flow diagram.
2. Ahh, that's a good point, I didn't think about that. I think that's a reasonably argument. On the other hand, the fees are already displayed above the flow diagram on the transaction page, so if you're interested in the fee it's already there. Also, having output 1 be the first thing on the flow diagram is nice, as opposed to the second after the fees.

**Reply by satoshiman** [reply]
(2) I also prefer that the paid fee put at last output on the flow chart. This makes it easier for people who have to work with rare sats to understand where their sats go (I think more people will collect sats in the near future). Personally, I was also confused a lot when I started getting used to this flowchart when I had to calculate each sats in the transaction.

Like this chart, I thought my rare sats (1st input) were transferred to miners
<img width="1128" alt="image" src="https://github.com/mempool/mempool/assets/37683529/5c1b3540-8e2c-4b6f-ab85-7b82a4eca4f4">
